File: Linked_List/reverse_stack.cpp
#include<bits/stdc++.h>
using namespace std;

// Linked List Node
struct Node{
    int data;
    struct Node* next;
    // 데이터 + 포인터 저장 
    Node(int data){
        this->data = data;
        next = NULL;
    }
};

// Linked List 
struct LinkedList{
    // Head 만들기
    Node* head;
    LinkedList(){
        head = NULL;
    }
    // print 
    void printList(){
        // Head 노드 저장 -> temp가 없을때까지 next && print
        struct Node* temp = head;
        while (temp != NULL){
            cout << temp->data << " ";
            temp = temp->next;
        }
    }
    // push : add front
    void push(int data){
        Node* temp = new Node(data);
        temp->next = head;
        head = temp;
    }

    void push_back(int data){
        Node* temp = new Node(data);
        
        if (head == NULL) head = temp;
        else{
            Node* last_node = head;
            while(last_node->next != NULL){
                last_node = last_node->next;
            }
            last_node->next = temp;
        }
    }

    // reverse with stack
    void reverseList(){
        // 스택생성 
        stack<Node*> S;
        Node* temp = head;
        // 리스트를 스택에 넣기
        while (temp->next != NULL){
            S.push(temp);
            temp = temp->next;
        }
        // 마지막 노드를 Head값으로 정의 
        head = temp;
        // 스택이 빌때까지 Top을 연결 
        while(!S.empty()){
            temp->next = S.top();
            S.pop();
            temp = temp->next;
        }
        temp->next = NULL;
    }
};

int main(){
    LinkedList LL;
    LL.push(1);
    LL.push(2);
    LL.push(3);
    LL.push_back(4);
    LL.printList();
    cout << "\n";
    LL.reverseList();
    LL.printList();
}

File: Linked_List/reverse_stack.py
# Node class
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

# Linked List class : push / push_back / printList / reverse
class LinkedList:
    def __init__(self):
        self.head = None 
    
    def printList(self):
        temp = self.head 
        while(temp):
            print(temp.data, end = " ")
            temp = temp.next 
    
    def push(self, data):
        new_node = Node(data)
        new_node.next = self.head 
        self.head = new_node 
    
    def push_back(self, data):
        new_node = Node(data)
        if self.head == None:
            self.head = new_node
        else:
            last_node = self.head
            while(last_node.next != None):
                last_node = last_node.next
            last_node.next = new_node
            
    def reverse(self):
        Stack = []  # stack list 
        temp = self.head 
        while(temp.next != None):
            Stack.append(temp)
            temp = temp.next 
        # 마지막 노드를 Head로 정의 
        self.head = temp 
        # 스택이 빌때까지 POP값을 연결 
        while(len(Stack) > 0):      # while(Stack)
            temp.next = Stack.pop()
            temp = temp.next 
        temp.next = None
            
LL = LinkedList()
LL.push(1);
LL.push(2);
LL.push(3);
LL.push_back(4);
LL.push_back(5);
LL.printList();
print("\n")
LL.reverse();
LL.printList();
